### DISCUSSION TITLE
UWP -> WinUI3: ViewModelLocator.Current

### DISCUSSION THREAD

**Discussion by myokeeh** [microsoft/TemplateStudio#4585] [Q&A] [answered]
In the WTS generated UWP app I'm converting to WinUI3/WindowsAppSDK, I call my method `SetBusy()` using `ViewModelLocator.Current.ShellViewModel.SetBusy(false);`

What is the equivalent to `ViewModelLocator.Current` to allow me to call `SetBusy()` in the same manner?

**Comment by myokeeh** [comment, marked as answer]
~~It appears to be `App.GetService<ShellPage>().ViewModel`~~

See https://github.com/microsoft/TemplateStudio/discussions/4585#discussioncomment-3330571

**Reply by mikebattista** [reply]
Where is the code that's calling SetBusy?

**Reply by myokeeh** [reply]
It's in several locations. Some in App.xaml.cs, code behind in various pages, and/or in the view models.

I've been able to get this to a point where it builds and runs (hitting breakpoints in `SetBusy()`), but it appears the bindings in my ShellPage XAML are not getting updated based on the values I set in `SetBusy()`

FYI, I'm realizing maybe the ViewModelLocator code I have is a remnant from one of the earlier conversions I did from one of the frameworks. I'm kind of starting from scratch with the latest WTS version just to have a clean slate and plan to copy over code to WinUI3-ify and WindowsAppSDK-ify it.

**Reply by mikebattista** [reply]
Pages and ViewModels are registered as transient services, so every time you call App.GetService you are creating a new instance. Interact with the instance of the ViewModel that is wired up to the displayed Page and bindings should work.

**Reply by myokeeh** [reply]
It appears bindings are working when called within the constructor. Bindings aren't updating if called elsewhere like in `App.GetService<ShellPage>().ViewModel.SetBusy(true, "test")` in App.xaml.cs. If I call `SetBusy(true, "yadayada")` in the constructor, bindings appear to update as expected. Any thoughts why the UI isn't updating when `SetBusy()` is called outside of ShellViewModel.cs?

ShellViewModel.cs:
```
    private string _text = string.Empty;
    private bool _visibility;

    public string Text1
    {
        get => _text;
        set
        {
            SetProperty(ref _text, value);
            //OnPropertyChanged(nameof(Text1));
        }

    }
    public bool Visibility1
    {
        get => _visibility;
        set
        {
            SetProperty(ref _visibility, value);
            //OnPropertyChanged(nameof(Visibility1));
        }

    }

    public ShellViewModel(INavigationService navigationService, INavigationViewService navigationViewService)
    {
        NavigationService = navigationService;
        NavigationService.Navigated += OnNavigated;
        NavigationViewService = navigationViewService;
        Text1 = "hello";
        Visibility1 = true;
        SetBusy(true, "yadayada"); //This also works. Updates the UI.
    }
    public void SetBusy(bool show, string text = "")
    {
        try
        {

            if (show)
            {
                Visibility1 = true; // Microsoft.UI.Xaml.Visibility.Visible;
                Text1 = text;
                //OnPropertyChanged(nameof(Visibility));
                //OnPropertyChanged(nameof(Text));
            }
            else
            {
                Visibility1 = false; // Microsoft.UI.Xaml.Visibility.Collapsed;
                //OnPropertyChanged(nameof(Visibility));
            }
            //await Task.CompletedTask;

        }
        catch (Exception)
        {

        }
    }
```

ShellPage.xaml
```
        <TextBlock Text="{x:Bind ViewModel.Text1, Mode=OneWay}"></TextBlock>
        <TextBlock Text="Hello" Visibility="{x:Bind ViewModel.Visibility1, Mode=OneWay}"></TextBlock>
```

**Reply by mikebattista** [reply]
See https://github.com/microsoft/TemplateStudio/discussions/4585#discussioncomment-3322099.

**Reply by myokeeh** [reply]
I see. I think I'm back to my original question. I can't grasp how to access the proper instance of the displayed page's ViewModel from App.xaml.cs.

**Reply by mikebattista** [reply]
App.xaml.cs could raise an event that the ViewModel registers for. The ViewModel could assign itself to a static property on the App class so that the App can reference the instance that way. You could try changing `AddTransient` to `AddSingleton` for the Page and ViewModel in App.xaml.cs and see if that helps with the App.GetService approach as well. These are just a few ideas without knowing much about your app's construction and layering.

**Reply by myokeeh** [reply]
I changed `services.AddTransient<ShellViewModel>();` to `services.AddSingleton<ShellViewModel>();` as you suggested.

Now `App.GetService<ShellViewModel>().SetBusy()` seems to address my initial question, though I don't know if this is the best thing to do. Time to learn more about what I just changed.

For now, I'm happy it's giving me the result I want.

Thanks!

**Reply by mikebattista** [reply]
Glad you got it working.

Service lifetime is explained at https://docs.microsoft.com/dotnet/core/extensions/dependency-injection#service-lifetimes.

**Reply by myokeeh** [reply]
@mikebattista I have forgotten about this thread and in my 2024 attempt to get back to using this, I ran into the same issue creating a new project from scratch.

Why is Transient the default in this case? It just seems like a wild-goose chase for someone trying to get started.
